external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala

```diff
@@ -0,0 +1,446 @@
+ *     batches, and cached KafkaConsumers in the executors can be reused efficiently. See the
+ *     docs on [[KafkaSourceRDD]] for more details.
+ */
+private[kafka010] case class KafkaSource(
+    sqlContext: SQLContext,
+    consumerStrategy: ConsumerStrategy[Array[Byte], Array[Byte]],
+    executorKafkaParams: ju.Map[String, Object],
+    sourceOptions: Map[String, String])
+  extends Source with Logging {
+
+  @transient private val consumer = consumerStrategy.createConsumer()
+  @transient private val sc = sqlContext.sparkContext
+  @transient private val initialPartitionOffsets = fetchPartitionOffsets(seekToLatest = false)
+  logInfo(s"Initial offsets: " + initialPartitionOffsets)
+
+  override def schema: StructType = KafkaSource.kafkaSchema
+
+  /** Returns the maximum available offset for this source. */
+  override def getOffset: Option[Offset] = {
+    val offset = KafkaSourceOffset(fetchPartitionOffsets(seekToLatest = true))
+    logInfo(s"GetOffset: $offset")
+    Some(offset)
+  }
+
+  /**
+   * Returns the data that is between the offsets [`start`, `end`), i.e. end is exclusive.
+   */
+  override def getBatch(start: Option[Offset], end: Offset): DataFrame = {
+    logDebug(s"GetBatch called with start = $start, end = $end")
+    val untilPartitionOffsets = KafkaSourceOffset.getPartitionOffsets(end)
+    val fromPartitionOffsets = start match {
+      case Some(prevBatchEndOffset) =>
+        KafkaSourceOffset.getPartitionOffsets(prevBatchEndOffset)
+      case None =>
+        initialPartitionOffsets
+    }
+
+    // Sort the partitions and current list of executors to consistently assign each partition
+    // to the executor. This allows cached KafkaConsumers in the executors to be re-used to
+    // read the same partition in every batch.
+    val topicPartitionOrdering = new Ordering[TopicPartition] {
+      override def compare(l: TopicPartition, r: TopicPartition): Int = {
+        implicitly[Ordering[(String, Long)]].compare(
+          (l.topic, l.partition),
+          (r.topic, r.partition))
+      }
+    }
+    val sortedTopicPartitions = untilPartitionOffsets.keySet.toSeq.sorted(topicPartitionOrdering)
+    val sortedExecutors = getSortedExecutorList(sc)
+    val numExecutors = sortedExecutors.size
+    logDebug("Sorted executors: " + sortedExecutors.mkString(", "))
+    val offsetRanges = sortedTopicPartitions.flatMap { tp =>
+      fromPartitionOffsets.get(tp).map { fromOffset =>
+        val untilOffset = untilPartitionOffsets(tp)
+        val preferredLoc = if (numExecutors > 0) {
+          Some(sortedExecutors(positiveMod(tp.hashCode, numExecutors)))
+        } else None
+        KafkaSourceRDD.OffsetRange(tp, fromOffset, untilOffset, preferredLoc)
+      }
+    }.toArray
+
+    // Create a RDD that reads from Kafka and get the (key, value) pair as byte arrays.
+    val rdd = new KafkaSourceRDD[Array[Byte], Array[Byte]](
```

Comment:
This should be `RDD[InternalRow]` otherwise we are copying and doing a lot of extra conversion.  It doesn't have to be this PR, but we might just want to do this right since I think I'll have a significant impact on performance.